Source/WebCore/platform/ScrollSnapAnimatorState.cpp

```diff
@@ -95,49 +95,81 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
     return offset * pageScale;
 }
 
-void ScrollSnapAnimatorState::setFocusedElementForAxis(ScrollEventAxis axis)
+// Returns whether the snap point is changed or not
+bool ScrollSnapAnimatorState::preserveCurrentTargetForAxis(ScrollEventAxis axis, ElementIdentifier boxID)
 {
     auto snapOffsets = snapOffsetsForAxis(axis);
-    auto found = std::find_if(snapOffsets.begin(), snapOffsets.end(), [](SnapOffset<LayoutUnit> p) -> bool {
-        return p.isFocused;
-    });
-    if (found == snapOffsets.end())
-        return;
-
-    auto newIndex = std::distance(snapOffsets.begin(), found);
-    auto newID = snapOffsets[newIndex].snapTargetID;
-    setActiveSnapIndexForAxis(axis, newIndex);
-    setActiveSnapIndexIDForAxis(axis, newID);
-}
-
-bool ScrollSnapAnimatorState::preserveCurrentTargetForAxis(ScrollEventAxis axis)
-{
-    auto snapID = activeSnapIDForAxis(axis);
-    auto snapOffsets = snapOffsetsForAxis(axis);
     
-    auto found = std::find_if(snapOffsets.begin(), snapOffsets.end(), [snapID](SnapOffset<LayoutUnit> p) -> bool {
-        return p.snapTargetID == *snapID;
+    auto found = std::find_if(snapOffsets.begin(), snapOffsets.end(), [boxID](SnapOffset<LayoutUnit> p) -> bool {
+        return p.snapTargetID == boxID;
     });
     if (found == snapOffsets.end()) {
-        setActiveSnapIndexIDForAxis(axis, std::nullopt);
+        setActiveSnapIndexForAxis(axis, std::nullopt);
         return false;
     }
     
     setActiveSnapIndexForAxis(axis, std::distance(snapOffsets.begin(), found));
     return true;
 }
 
-bool ScrollSnapAnimatorState::resnapAfterLayout(ScrollOffset scrollOffset, const ScrollExtents& scrollExtents, float pageScale)
+Vector<SnapOffset<LayoutUnit>> ScrollSnapAnimatorState::currentlySnappedOffsetsForAxis(ScrollEventAxis axis) const
 {
-    // Check if we need to set the active target to a focused element
-    setFocusedElementForAxis(ScrollEventAxis::Vertical);
-    setFocusedElementForAxis(ScrollEventAxis::Horizontal);
+    Vector<SnapOffset<LayoutUnit>> currentlySnappedOffsets;
+    auto snapOffsets = snapOffsetsForAxis(axis);
+    auto activeIndex = activeSnapIndexForAxis(axis);
+    
+    if (activeIndex && *activeIndex < snapOffsets.size())
+        currentlySnappedOffsets.append(snapOffsets[*activeIndex]);
+    return currentlySnappedOffsets;
+}
+
+static bool isSnappedToMultipleBoxes(const Vector<SnapOffset<LayoutUnit>>& verticalOffsets, const Vector<SnapOffset<LayoutUnit>>& horizontalOffsets)
+{
+    HashSet<ElementIdentifier> boxIDs;
+    
+    Vector<SnapOffset<LayoutUnit>> currentlySnappedOffsets;
+    currentlySnappedOffsets.appendVector(verticalOffsets);
+    currentlySnappedOffsets.appendVector(horizontalOffsets);
     
+    bool hasMultipleSnapAreas = false;
+    for (auto offset : currentlySnappedOffsets) {
+        boxIDs.add(offset.snapTargetID);
+        hasMultipleSnapAreas |= offset.snapAreaIndices.size() > 1;
+    }
+    return boxIDs.size() > 1 || hasMultipleSnapAreas;
```

Comment:
I think this would be more efficient if you just looped over the X and Y offsets in two loops, and early returned as soon as you find the exit condition.

Reply:
Fixed.